[cudagraph] improve log for mutating static input tensor addresses

BoyuanFeng opened the pull request:
Summary: This diff adds more log for cudagraph when static input tensor mutates. For each placeholder whose static input tensor address mutates, we log its name, changed data pointer address, and the input stack trace. Since some placeholder may have empty stack trace, we find its first user with an non-empty stack trace and print this stack trace instead.

Test Plan: buck2 run fbcode//caffe2/test/inductor:cudagraph_trees -- --r test_static_inputs_address_mutation_log

Differential Revision: D57805118




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -960,11 +960,19 @@ def check_static_inputs_are_stable(self, new_inputs):
+                        f"input name: {self.wrapped_function.placeholders[index]}. "
+                        "data pointer changed from {data_ptr} to {t.data_ptr()}\n"
+                    )
+            error_msg += (
```
nit: if you're already using f-streams, just do `f"{error_msg} related graph: ...`

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -960,11 +960,19 @@ def check_static_inputs_are_stable(self, new_inputs):
+                index = self.non_managed_static_input_idxs[i]
+                if t.data_ptr() != data_ptr:
+                    error_msg += (
+                        f"input name: {self.wrapped_function.placeholders[index]}. "
```
Our other error msgs use the original source code. I think that is more useful to the user than our aot_autograd generated input name/stack trace. We could augment it with graph name if you think it is helpful.

What do you htink ?

eellison reviewed (commented):
Prior error msgs used source code here we're using graph name

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -960,11 +960,19 @@ def check_static_inputs_are_stable(self, new_inputs):
+            error_msg += (
+                f"related graph: {self.wrapped_function.placeholders[0].graph}\n"
+            )
+            error_msg += f"related stack traces:{self.stack_traces}\n"
```
These are output stack traces. we want input stack traces. also, we shouldn't dump every stack trace - this will be a gigantic error msg. we should only include stack trace for input that actually error'd

BoyuanFeng committed:
[cudagraph] improve log for mutating static input tensor addresses (#127145)

Summary:

This diff adds more log for cudagraph when static input tensor address mutates.

For each placeholder whose static input tensor address mutates, we log its name, changed data pointer address, and the input stack trace. Since some placeholder may have empty stack trace, we find its first user with an non-empty stack trace and print this stack trace instead.

Test Plan: buck2 run fbcode//caffe2/test/inductor:cudagraph_trees -- --r test_static_inputs_address_mutation_log

Reviewed By: eellison

Differential Revision: D57805118
